contracts/provider/assembly/index.ts

```diff
@@ -0,0 +1,216 @@
+}
+
+/**
+ * The off-chain upload/status RESTful API endpoint supported by this provider.
+ */
+export function apiEndpoint(): string {
+  return _apiEndpoint();
+}
+
+function _apiEndpoint(): string {
+  const p = storage.getString("_apiEndpoint");
+  return p ? p : "";
+}
+
+export function setApiEndpoint(a: string): void {
+  onlyOwner();
+  storage.setString("_apiEndpoint", a);
+}
+
+/**
+ * Logging event fired when a deposit has been added.
+ */
+function DepositAdded(
+  depositee: string,
+  depositor: string,
+  amount: u128
+): void {
+  logging.log(
+    `DepositAdded("${depositee}", "${depositor}", "${amount.toString()}")`
+  );
+}
+
+/**
+ * Logging event fired when a deposit has been released.
+ */
+function DepositReleased(
+  depositee: string,
+  depositor: string,
+  amount: u128
+): void {
+  logging.log(
+    `DepositReleased("${depositee}", "${depositor}", "${amount.toString()}")`
+  );
+}
+
+const KEY_INITIATED = "initiated";
+
+export function initialize(
+  providerProportion: f64,
+  sessionDivisor: u128,
+  apiEndpoint: string
+): void {
+  onlyOwner();
+  if (storage.getPrimitive<bool>(KEY_INITIATED, false)) {
+    return;
+  }
+  setProviderProportion(providerProportion);
+  setSessionDivisor(sessionDivisor);
+  setApiEndpoint(apiEndpoint);
+  storage.set<bool>(KEY_INITIATED, true);
+}
+
+// Public Methods
+
+/**
+ * Deposit attached funds with this provider for the given account to initiate a session.
+ * Use `attachedDeposit` to attach a deposit in yocto-Ⓝ that will produce a session proptional
+ * in length to `sessionDivisor`.
+ */
+export function addDeposit(depositee: string): void {
+  assert(
+    context.attachedDeposit > u128.Zero,
+    "BridgeProvider: must include deposit > 0"
+  );
+  if (deposits.contains(depositee)) {
+    // Maybe it has expired?
+    _releaseDeposit(depositee);
+    assert(
+      !deposits.contains(depositee),
+      "BridgeProvider: depositee already has deposit"
+    );
+  }
+  deposits.set(
+    depositee,
+    new Deposit(context.blockTimestamp, context.sender, context.attachedDeposit)
+  );
+  DepositAdded(depositee, context.sender, context.attachedDeposit);
+}
+
+function isDepositValid(d: Deposit | null, t: u64, div: u128): bool {
+  return (
+    d != null &&
+    d.value > u128.Zero &&
+    t <= d.timestamp + u128.div(d.value, div).toI64()
+  );
+}
+
+/**
+ * Return whether the given depositee has funds deposited with this provider.
+ */
+export function hasDeposit(depositee: string): bool {
+  return isDepositValid(
+    deposits.get(depositee),
+    context.blockTimestamp,
+    _sessionDivisor()
+  );
+}
+
+function _releaseDeposit(depositee: string): void {
+  const deposit = deposits.get(depositee);
+  if (deposit == null) return;
+  if (
+    deposit.value > u128.Zero &&
+    !isDepositValid(deposit, context.blockTimestamp, _sessionDivisor())
+  ) {
+    let value = deposit.value;
+    const cut = u128.fromF64(f64.mul(value.toF64(), _providerProportion()));
```

Comment:
If you can think of ways to avoid all these conversions, I'm all ears, this doesn't seem ideal, but proportions kinda dictate it. The "scary" thing is that there is a possibility for rounding errors and integer conversion issues.